### DISCUSSION TITLE
data virtualization ?

### DISCUSSION THREAD

**Discussion by issam1975** [IgorBuchelnikov/ObservableComputations#11] [Q&A] [answered]
Hi,

is it possible to use this library to do data virtualization  with entyity framework ?

any example or guidance on how to to achive this is much appreciated .

thanks

**Comment by IgorBuchelnikov** [comment]
Explain what is data virtualization with entity framework? I only know about UI data virtualization. For example 

https://docs.telerik.com/devtools/wpf/controls/radgridview/features/ui-virtualization

https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/listview-and-gridview-data-optimization

**Reply by issam1975** [reply]
i mean load data on demand for example when scrolling a datagrid or a listbox 
and under the hood call entity framework skip / take method .

the link you have provided seems to do what i want but unfortunally it's for UWP and i am using (WPF) 
so i am affraid that's it's not available  for it .

sorry for not mentionning this

**Reply by IgorBuchelnikov** [reply]
Now I understand what you mean.

**Comment by IgorBuchelnikov** [comment, marked as answer]
Have you seen [this](https://www.codeproject.com/Articles/34405/WPF-Data-Virtualization) article? The article contains a WPF project with data and UI virtualization. Maybe this project is all you need. The only thing you should do is change the project so that it loads data using Entity Framework. Do you need more than the project provides? 

The primary use case of ObservableComputations is the processing of changes in already loaded data. The processing may lead to other data loading, but this is secondary. The way of data loading is not an area of responsibility of ObservableComputations. The data loading can be done with Entity Framework or other ORM, synchronously or asynchronously. Data virtualization is more refers to a way of data loading.

**Comment by issam1975** [comment]
indeed this  is what i ended up with .

i have modified the code in the article you have mentionned to load data from efcore , and it seems to works correctly for now ..... :)

back to ObservableComputations, if i have understood the purpose of the library (please correct me if i am wrong )

is to do automatic  refresh  of linq queries when the source of the data has changed .

i am right ?


ps : Thanks a lot, Really appreciate !

**Reply by IgorBuchelnikov** [reply]
You are right. Some clarifications:
BCL contains two kinds of LINQs: 

1. LINQ to [IEnumerable&lt;T&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-5.0)
2. LINQ to [IQuerable&lt;T&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.linq.iqueryable-1?view=net-5.0)

[Select() to IEnumerable&lt;T&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.select?view=net-5.0#System_Linq_Enumerable_Select__2_System_Collections_Generic_IEnumerable___0__System_Func___0___1__) and [Select() to IQuerable&lt;T&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.select?view=net-5.0#System_Linq_Queryable_Select__2_System_Linq_IQueryable___0__System_Linq_Expressions_Expression_System_Func___0___1___) are two different method.

Entity framework's [DbSet&lt;TEntity&gt;](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbset-1?view=efcore-5.0) implements [IQuerable&lt;TEntity&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.linq.iqueryable-1?view=net-5.0). 
[DbSet&lt;TEntity&gt;.Local](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbset-1.local?view=efcore-5.0) implements [IEnumerable&lt;TEntity&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-5.0) and [INotifyCollectionChanged ](https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized.inotifycollectionchanged?view=net-5.0) .

Observableomputations library is a kind of LINQ to [IEnumerable&lt;T&gt;](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-5.0) and [INotifyCollectionChanged ](https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized.inotifycollectionchanged?view=net-5.0).

Additionally, ObservableComputations allow automatic refresh of an arbitrary expression.
